decisions/DR007-Hierarchical-or-compositional-traits-for-specifications.md

```diff
@@ -0,0 +1,345 @@
+
+#### Option 1
+
+```
+specification blob {
+  mimeType: string
+}
+
+specification blob.cache {
+  size: int
+  startFrame: float
+  endFrame: float
+}
+
+// GeometryCacheSpecification
+specification blob.cache.geometry {
+  up_axis: str
+}
+
+specification blob.cache.raster {
+  width: int
+  height: int
+}
+
+// ImageSpecification
+specification blob.cache.raster.image {
+  channels:int
+}
+
+specification container {
+}
+
+// ShotSpecification
+specification container.shot {
+  startFrame: float
+  endFrame: float
+  
+  width: int
+  height: int
+}
+
+```
+
+#### Option 2
+
+```
+trait blob {
+  mimeType: string
+}
+
+trait raster {
+  width: int
+  height: int
+}
+
+trait frames {
+  startFrame: float
+  endFrame: float
+}
+
+trait geometry {
+  up_axis: str
+}
+
+trait cache {
+  size: int
+}
+
+---
+
+specification Shot {
+  traits: frames
+}
+
+specification Image {
+  traits: blob, raster
+}
+
+specification GeometryCache {
+  traits: blob, geometry, frames
+}
+
+specification ImageCache {
+  traits: blob, raster, frames
+}
+
+
+```
+
+### Checking management policies
+
+#### Option 1
+
+When using management interface functions the code using specifications
+is the same for Option 1 and Option 2.
+
+```python
+# uses traits from "An implementation sketch of Option 2"
+policy = manager.managementPolicy([ImageSpecification()], context)
+
+if policy == [kIgnored]:
+    pass  # nothing with an ImageSpecification is managed
+
+# check if pngs and alembic caches are supported by the manager
+exrSpec = ImageSpecification()
+exrSpec.mimeType = "image/x-exr; version=\"2\""
+
+abcSpec = GeometryCacheSpecification()
+abcSpec.mimeType = "alembic"
+policy = manager.managementPolicy([exrSpec, abcSpec], context)
+
+# manager offers to manage these exrs and alembic
+policy == [kManaged, kManaged] 
+```
+
+#### Option 2
+
+```python
+# uses traits from "An implementation sketch of Option 2"
+policy = manager.managementPolicy([ImageTrait()], context)
+
+if policy == [kIgnored]:
+    pass  # nothing with an image trait is managed
+
+# check if only a certain extension is supported by the manager
+blobTrait1 = BlobTrait()
+blobTrait.mimeType = "image/x-exr; version=\"2\""
+
+blobTrait2 = BlobTrait()
+blobTrait2.mimeType = "alembic"
+
+policy = manager.managementPolicy([blobTrait, blobTrait2], context)
+
+policy == [kManaged,
+           kManaged]  # manager offers to manage these exrs and alembic
+```
+
+### Finding all the caches associated with a shot
+
+The resulting entity references could be used to help clear up space on
+disk when they're no longer necessary for work.
+
+Assumes the following set up:
+
+```python
+assets = ["asset://fxPointsCache",  # a geometry cache of point data
+          "asset://precompShadows"]  # rendered image of shadows
+
+shots = ["shot://ABC/01"]
+
+```
+
+#### Option 1
+
+```python
+manager.getRelatedReferences(shots, [], context, session,
+                             resultSpec=[ImageSpecification(),
```

Comment:
[fix] If the Specifications are as described above, then for Option 1 the equivalent would be this, rather than using the two individual specifications:

`manager.getRelatedReferences(shots, [BlobCacheSpecification()], context, session)`

Reply:
Addressed in [78c87c0d7038ae6c2f948e0d7ed5bc39023d108f](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/238/commits/78c87c0d7038ae6c2f948e0d7ed5bc39023d108f)